### PR TITLE
Add Created and Deleting events

### LINKED ISSUE #2551: [Flags] Create an event for flag creation
Currently, the `flarum/flags` extension only has an event for when a flag is being deleted ([`FlagsWillBeDeleted`](https://github.com/flarum/flags/blob/master/src/Event/FlagsWillBeDeleted.php)).

It would be useful to have an event for when a flag is created as well (more precisely, when a user flags a post).

A possible use-case would be, for example, an extension that generates an "external" notification for admins/moderators when a flag is created. (With external I mean an email or a message in a Telegram group, for example.)

Relevant discussion on discuss: https://discuss.flarum.org/d/25981-listening-to-flag-creation

### PR DESCRIPTION
**Pull request by imorland**
Fixes flarum/core#2551

- Create two new events: `Created` and `Deleting`
- Deprecate `FlagsWillBeDeleted`

Argument for single flag per `Deleting` event:  As the `Created` event is for a single new flag, I decided it would make the most sense to also handle `Deleting` in the same, single flag at a time way. Would welcome any thoughts on how this is implemented.

### CODE REVIEW

File: src/Event/Created.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    /**
+     * @var User
+     */
+    public $actor;
+
+    /**
+     * @param Flag $flag
+     * @param User $actor
+     */
+    public function __construct(Flag $flag, User $actor)
```
Comment (askvortsov1) [inline src/Event/Created.php]: Why remove $data from these?

File: src/Command/DeleteFlagsHandler.php [hunk 2]
```diff
@@ -48,8 +49,13 @@ public function handle(DeleteFlags $command)
 
         $actor->assertCan('viewFlags', $post->discussion);
 
+        // remove beta 17
         $this->events->dispatch(new FlagsWillBeDeleted($post, $actor, $command->data));
 
+        foreach ($post->flags as $flag) {
+            $this->events->dispatch(new Deleting($flag, $actor));
```
Comment (askvortsov1) [inline src/Command/DeleteFlagsHandler.php]: I suppose this depends on the use case for listening to the event: would each flag be handled individually? Generally speaking, it feels like the number of flagged posts is more relevant than the actual number of flags. And flags can't really be dismissed individually, which might be a possible feature expansion, but right now, it doesn't seem like individual de-flagging is that applicable.

Do we have any extensions using these / ideas of use case?

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

File: src/Command/DeleteFlagsHandler.php (hunk 2 above)
Comment (imorland) [reply src/Command/DeleteFlagsHandler.php]: My example use case for individual deleted flags is this (which I also hope to build as an extension if this PR is accepted):

Each `Created` event is captured and stored someplace. To complete the flag lifecycle from creation to closure, an individual `Delet(ing/ed)` matching event would be required, rather than a general 'all flags for this post were cleared'. 

Take the general concept of my previous flag dismiss logging. Obviously that would track Mods (or whoever was granted permissions) to dismiss flags and log/keep records of those actions for audit trail purposes. Go one step further than that, say you had a need to build some statistical analysis patterns on how many flags a user has created, or to build up some ML over which type of posts get flagged by which user type, or simply to build an average on how long flags remain open before being resolved.

Long story short, the beginning and end events of an action should match, I believe.

File: src/Event/Created.php (hunk 1 above)
Comment (imorland) [reply src/Event/Created.php]: Erm, I honestly don't know 🙈 Over-eager tidyup, I guess. I'll reintroduce

**Review by askvortsov1 — APPROVED** [review]

**Review by tankerkiller125 — APPROVED** [review]
